### PR TITLE
Show non-PgError errors when connecting to db

### PR DESCRIPTION
**Pull request by slai**
# TL;DR
When network (or other non-database) errors occur while connecting to the database, the error is masked by a panic due to a bug in unwrapping the error.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

When connecting to a database instance that doesn't exist, that error is currently masked due to an invalid assumption that all wrapped errors returned from gorm are of type `*pgconn.PgError`. When a network error occurs, the returned error is of type `*net.OpError`. This causes the following panic -

```
{"json":{"src":"base.go:73"},"level":"fatal","msg":"caught panic: interface conversion: error is *net.OpError, not *pgconn.PgError [goroutine 1 [running]:\nruntime/debug.Stack()\n\t/home/user/.local/go/src/runtime/debug/stack.go:24
...
```

... and doesn't show what the actual error is.

This PR fixes that by using the functions in the `errors` package to check if the error chain contains an error of type `*pgconn.PgError`, and if not, simply returns the error.

It also removes the ConnectError interface, as it is no longer needed - the `errors.As` function will automatically check the whole error chain and unwrap as necessary.

With this fix, the error becomes -

```
{"json":{"src":"base.go:82"},"level":"fatal","msg":"failed to connect to `host=localhost user=postgres database=flyte`: dial error (dial tcp 127.0.0.1:5432: connect: connection refused)","ts":"2022-06-11T20:48:28+01:00"}
```

## Tracking Issue
N/A

## Follow-up issue
N/A

### CODE REVIEW

File: pkg/repositories/database.go [hunk 1]
```diff
@@ -116,17 +115,17 @@ func createPostgresDbIfNotExists(ctx context.Context, gormConfig *gorm.Config, p
...
 	}
 
-	// Check if its invalid db code error
-	cErr, ok := err.(repoErrors.ConnectError)
-	if !ok {
-		logger.Errorf(ctx, "Failed to cast error of type: %v, err: %v", reflect.TypeOf(err),
-			err)
+	// if it is an invalid db code error, continue, otherwise return
+	pgErr := &pgconn.PgError{}
+	if !errors.As(err, &pgErr) {
```
Comment (pmahindrakar-oss) [inline pkg/repositories/database.go:120]: Nit: Can you add a few unit tests to cover the newly added lines

**Review by pmahindrakar-oss — DISMISSED** [review]

File: pkg/repositories/database.go (hunk 1 above)
Comment (slai) [reply pkg/repositories/database.go:120]: It's pretty hard to write unit tests without refactoring all of this, because this function potentially opens multiple connections to Postgres directly in its code.

To do this, I'm thinking of creating a DialectorBuilder function/struct that takes a DSN and returns a Dialector, and pass it to `createPostgresDbIfNotExists` as an arg. DialectorBuilder would be created in the `GetDB` function (or further up). I would also need to create a mock DialectorBuilder which will return a mock Dialector that returns the configured errors.

Alternatively, I could just extract these few error handling lines of code and write some tests for it, but that feels like breaking up the code just for the sake of writing tests for specific lines, when the whole function should be tested.

(Ideally, I don't think flyteadmin should even have the permission to `CREATE DATABASE` - that should be a db admin only permission. But this is a much bigger change across the board.) 

Looking at the existing tests, I don't think there's any coverage here at the moment.

@pmahindrakar-oss any thoughts or other ideas? Would you like me to refactor this to include unit tests?

File: pkg/repositories/database.go [hunk 2]
```diff
@@ -112,20 +111,14 @@ func createPostgresDbIfNotExists(ctx context.Context, gormConfig *gorm.Config, p
 
 	dialector := postgres.Open(getPostgresDsn(ctx, pgConfig))
 	gormDb, err := gorm.Open(dialector, gormConfig)
-	if err == nil {
```
Comment (slai) [inline pkg/repositories/database.go:115]: I changed this so it checks if it is an error first. Even though the logic is the same as before, I feel checking `if err != nil` immediately after the function creating the error is safer/more natural because this is the general Go error checking pattern.

Comment (slai) [reply pkg/repositories/database.go]: Sorry, ignore this - I changed it back. On second thought, the quirk in flow this introduces means it isn't really much cleaner than the original code.

File: pkg/repositories/database.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply pkg/repositories/database.go]: Ok sounds good. Thanks @slai

**Review by pmahindrakar-oss — DISMISSED** [review]

**Comment by pmahindrakar-oss** [comment]
@slai  Can you take care of the DCO issue.

**Review by pmahindrakar-oss — APPROVED** [review]
